### ISSUE TITLE
Signature verification only works with rsa key

### ISSUE THREAD

**Issue by jpevarnek** [paramiko/paramiko#663]
Using the following setup:

```
client = paramiko.SSHClient()
host_keys = client.get_host_keys()
host_keys.clear()
public_key = ecdsakey.ECDSAKey(data={binary_key_data})
host_keys.add({server_ip}, "ecdsa-sha2-nistp256", public_key)

authentication_key = {code to get private key (generate fake key file, read it in)}

client.connect({server_ip}, username='root', pkey=authentication_key, allow_agent=False, look_for_keys=False)
```

This gives the error "Server '{server_ip}' not found in known_hosts"

Switching to using the rsa public key for the server works fine.

My understanding of the problem:
In `connect`, we get the server key through `t.get_remote_server_key()` where `t` is the transport.  We also get the type for that key (`server_key.get_name()`).  This gets the host key that we identified in the transport and its type.  This seems to be a problem as, when we set `host_key_type` in the transport (`_parse_kex_init`), it's arbitrarily choosing among the list of what is supported by the server and seems to be defaulting to "ssh-rsa" when all three are supported.

I honestly don't know enough about the SSH protocol to suggest a fix here, however, this does seem like it should work to me.  I'd be happy to work on this if there's anything I can do to help out.

**Comment by bitprophet** [comment]
Thanks for the report! I have a suspicion that this duplicates one of the other issues in the Keys label, but flagging it with that and leaving open for now, just in case. (Feel free to search through those to confirm this if you've the time!)

tl;dr that chunk of `connect` and friends which deals with key/cipher/etc selection, are due for a major overhaul.

**Comment by phihag** [comment]
I am seeing the exact same problem.

Luckily, it has already been solved in #473. Applying #473 makes paramiko work for me!

**Comment by bitprophet** [comment]
Thanks for catching that, @phihag !
